Caltech-IPAC/firefly
# Firefly 1150 object id tap

#### [Firefly-1150](https://jira.ipac.caltech.edu/browse/FIREFLY-1150)
- Added a new Search Helper (ObjectID Search) to the TAP panel under services that support uploads 
- allow user to upload a table and select a column from their uploaded table, and a column from the selected table on the right. This will lead to a new ADQL constraints of the format '`ut.objectID = a.objectID`' (ut = uploaded table, a = first char of the selected table in this case) 
- This is a draft PR because I need to fix one small style issue and at possibly this bug below:
- Bug Description 
  - This is a regression bug. Switching back from **multi-object** to **Single Shape** leads to some incorrect ADQL. For a service, like GAIA (and table is "public.hipparcos"), if under **Spatial** you select **Multi-object** and upload a table. Now if you select **Single Shape** and enter in, say, M31, and click on **Populate and Edit ADQL**. Notice that the **WHERE** clause includes "p.ra" and "p.de", even though the **FROM** clause is just: `FROM public.hipparcos` and not `FROM public.hipparcos as p` (as it would be if there were an uploaded table). 
      - If you try to do a search now, you'd get an error which would include "`Unknown column "p.ra"`" and "`Unknown column "p.de"`"...
      - for some reason, this same behavior does not affect some of the tables under IRSA/CADC. So will need to test this further, but the basic idea of the fix is that if you switch from multi-object to single shape and there's no uploaded table for that search, then "**p.ra**" and "**p.de**" should simply be **ra** and **de** 

**Updates 7/6**: 
   - The Temporal Column and Spatial Type bugs have been fixed
   - For bug description, see @loitly's main comment on this PR below 
       - **Spatial Search bug**
          - multi-object selection was very persistent. This is fixed now by commenting out a use-effect (I don't think there's any need for it anymore, but commented out just so reviewers can take a look at it). Whatever Spatial Type you select now (single shape or multi-object), the selection is respected when you switch services or go to `Edit ADQL` and back to `UI Assisted` mode.  
          - the **_exception_** to this bug is, if you do a whole table search (or "search TAP for row" from the table), then the useEffect that gets triggered (line 103 in `SpatialSearch.jsx`) has ones of its dependencies: _searchParams.uploadInfo_. _searchParams_ comes, as part of _initArgs_, all the way from defaultSearchActions.js (see the `searchWholeTable` function in this file as an example). This useEffect sets `SPATIAL_TYPE` to `MULTI`, which is good the first time a user does a search from a table. But after that, user may switch to single shape and go to `Edit ADQL` or a different service. And if you go back to the previous service or `UI Assisted` mode, multi-object is selected again. I am not really sure what a good fix for this is. 
      - **Temporal Search bug**
        - see my comment below for explanation of this commit 
  -  test build is updated, feel free to try it out!

**Updates 7/10**: 
- there will still some issues with Temporal and Spatial search bugs, as described by @loitly below. These are all fixed now and the test build is updated to reflect these changes. 
- Also includes @robyww's commit to fix `Edit ADQL` and make it sticky (if you select `Edit ADQL`, do a search and click on `TAP` again, it should take you to the `Edit ADQL` mode and not the UI Assisted mode). 

**Updates 7/10 Part 2**:
  - The UI changes to ObjectID Search Helper suggested by @robyww are in. Object ID Search helper is selected when you interact with it (upload a table, or change the column). And the Object ID columns are now open by default as well. 
  - There were a couple of bugs I noticed: The `Uploaded Object ID` field was not validating column entries. And the `Object ID (from the selected table on the right)` field was not updating its columnsModel for the column validator when you switched services/tables correctly. 
  - Fixed these and now the Object ID Search Helper is fully functional and has expected behavior (and bug free, from what I have tested). 

**Updates 7/11**:
- got some of the final PR comments in, involves some code cleanup. Using JSdoc/typdef, and simplifying a for loop to use filter instead, amongst other changes. 

**Testing**: 
- https://fireflydev.ipac.caltech.edu/firefly-1150-objectid-tap/firefly
- for any service that supports uploads (IRSA, CADC, etc.), upload a table using the ObjectID Search helper in the TAP panel and select an ObjecyID column for your uploaded table and for the selected table on the right. For testing purposes, this could be any column you want. 
  - Click on **Populate and edit ADQL**. Make sure the ADQL looks correct. 
  - Ideally, select a service and table that is not too big. This is an issue i've been running into for testing. Because of the `WHERE ut.objectID = a.objectID` type of query, I think every row of the selected table on the right is being checked, which makes sense. So even if you add a row limit, if the table is big enough, the search will be slow. I tried the **CADC's ivoa.ObsCore** table (which has  166,523 rows) and so far that gave me the fastest results... 
- Seelct IRSA as the TAP service. Upload a table under the **Spatial** search helper, and also the **ObjectID Search** helper (make sure both of these are selected as well). Click on **Populate and edit ADQL**. This should give you an error, because we don't support multiple uploads at this time. 
- switch to a service like NED, make sure the **ObjectID Search** helper doesn't show up here
